servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherCache.java

```diff
@@ -0,0 +1,261 @@
+     * @param publisherSupplier if the key does not exist in the cache, used to create a new publisher.
+     * @return a new {@link Publisher} from the publisherSupplier if not contained in the cache, otherwise
+     * the cached publisher.
+     */
+    public Publisher<T> get(final K key, final Function<K, Publisher<T>> publisherSupplier) {
+        return Publisher.defer(() -> {
+            synchronized (publisherCache) {
+                if (publisherCache.containsKey(key)) {
+                    return publisherCache.get(key).publisher;
+                }
+
+                final Holder<T> item2 = new Holder<>();
+                publisherCache.put(key, item2);
+
+                final Publisher<T> newPublisher = publisherSupplier.apply(key)
+                        .liftSync(subscriber -> new Subscriber<T>() {
+                            @Override
+                            public void onSubscribe(Subscription subscription) {
+                                subscriber.onSubscribe(new Subscription() {
+                                    @Override
+                                    public void request(long n) {
+                                        subscription.request(n);
+                                    }
+
+                                    @Override
+                                    public void cancel() {
+                                        try {
+                                            assert Thread.holdsLock(publisherCache);
+                                            publisherCache.remove(key, item2);
+                                        } finally {
+                                            subscription.cancel();
+                                        }
+                                    }
+                                });
+                            }
+
+                            @Override
+                            public void onNext(@Nullable T next) {
+                                subscriber.onNext(next);
+                            }
+
+                            @Override
+                            public void onError(Throwable t) {
+                                lockRemoveFromMap();
+                                subscriber.onError(t);
+                            }
+
+                            @Override
+                            public void onComplete() {
+                                lockRemoveFromMap();
+                                subscriber.onComplete();
+                            }
+
+                            private void lockRemoveFromMap() {
+                                synchronized (publisherCache) {
+                                    publisherCache.remove(key, item2);
+                                }
+                            }
+                        });
+
+                item2.publisher = multicastStrategy.apply(newPublisher)
+                        .liftSync(subscriber -> new Subscriber<T>() {
```

Comment:
Do we need this second `liftSync`? I think it applies to the individual streams. In my minds eye that means if a single stream ends it removes the underlying stream from the cache. I don't know how that would happen unless the parent publisher completed and we remove it at that level as well.

Reply:
It's there to handle subscriber errors and clean up if the subscriber is put into a bad state via a throw. In fact it looks like I missed a case with onNext.

Reply:
Is there a test case that would demonstrate their purpose? If I delete the second liftsync and move the syncrhronized to the first everything still works as expected.

Reply:
I will add one.

Reply:
> Also consider breaking this out into a named (e.g. not anonymous, private/final) class which is easier to look at when debugging larger operator chains.

These are inline because they access four different pieces of local state and properties. I played around with breaking this out and it doesn't feel cohesive. I'm choosing to leave it as is for now but I am happy to revisit if you have strong feelings.